kuma/javascript/src/newsletter.jsx

```diff
@@ -89,76 +91,69 @@ export default function Newsletter() {
+        event.preventDefault();
+        let params = new URLSearchParams(
+            new FormData(newsletterForm)
+        ).toString();
+        fetch(newsletterSubscribeURL, {
+            method: 'POST',
+            headers: {
+                'X-Requested-With': 'XMLHttpRequest',
+                'Content-type': 'application/x-www-form-urlencoded'
+            },
+            body: params
+        })
+            .then(response => response.json())
+            .then(({ success, errors }) => {
+                if (success) {
+                    permanentlyHideNewsletter();
+                    setShowSuccessfulSubscription(true);
+                    ga('send', {
+                        hitType: 'event',
+                        eventCategory: 'newsletter',
+                        eventAction: 'progression',
+                        eventLabel: 'complete'
+                    });
+                    return;
+                }
 
-                        // if there are erros and the array contain at least one item
-                        if (errorsArray && errorsArray.length) {
-                            setErrorsArray(errorsArray);
-                            // if there was an error, but there are no items in the array
-                        } else if (errorsArray && errorsArray.length === 0) {
-                            if (newsletterForm) {
-                                // set the skip-fetch data attribute on the form
-                                newsletterForm.dataset.skipFetch = 'true';
-                                // and submit again for diagnoses on the server
-                                newsletterForm.submit();
-                            }
-                        } else {
-                            setShowNewsletter(false);
-                            permanentlyHideNewsletter();
-                            setShowSuccessfulSubscription(true);
-                            ga('send', {
-                                hitType: 'event',
-                                eventCategory: 'newsletter',
-                                eventAction: 'progression',
-                                eventLabel: 'complete'
-                            });
-                        }
-                    }
-                });
+                if (errors && errors.length) {
+                    setError(errors);
+                    // if there was an error, but there are no items in the array
+                } else if (errors && errors.length === 0) {
+                    // set the skip-fetch data attribute on the form
+                    newsletterForm.dataset.skipFetch = 'true';
+                    // and submit again for diagnoses on the server
+                    newsletterForm.submit();
+                }
+            })
+            .catch(e => {
+                console.error('error while subscribing to newsletter', e);
+                setError(['An error occurred while submitting the form']);
```

Comment:
I don't know what you need to pay Flow to make it work but I think pretty print the error *here* isn't ideal. 
I would do:
```javascript
setErrors([e])
```

And then, somewhere in the "display part" of the error I'd do something like this:
```javascript
{errors.map(error => {
  <li key={error.toString()}>
    {error instanceof Error ? gettext('Network error bla bla')  : error}
  </li>
})
```
or something fancier.

Reply:
Fair enough, I changed it to show the stringified error!